### DISCUSSION TITLE
How best to add this feature: insert macro before each function definition

### DISCUSSION THREAD

**Discussion by pixsperdavid** [getditto/safer_ffi#202] [General] [answered]
I'm currently trying to implement features necessary to generate a C header file which can be used to automatically add `__declspec(dllimport)` before each function definition when used in a Windows, which is necessary in order to use functions from a DLL. The first part of this was implementing in PR https://github.com/getditto/safer_ffi/pull/199, which allows writing arbitrary text after the include guard in order to define the import macro. 

The second part is allowing the defined import macro to be placed before each function definition. I'm slightly struggling on implementing this in a sensible way currently without adding an extra parameter to the `emit_function` function in the `HeaderLanguage` trait. This isn't ideal as the option would only ever be relevant to C language generation.

Would welcome thoughts on the best way to implement this. It seems to me there is a need for something like a language specific `GenerationConfig` struct for language specific options as part of the `HeaderLanguage` trait perhaps?

**Comment by pixsperdavid** [comment]
If there are no other suggestions, I'm going to draft a PR introducing a language specific `GenerationConfig` as mentioned above as a path to implementing this feature.

**Reply by danielhenrymantilla** [reply, marked as answer]
Apologies, I have been (and kind of still am) quite busy, so I cannot necessarily dedicate enough time to think about the best way to achieve these things. However, what you suggest does sound sensible, especially off the shoulders of [`Definer`](https://docs.rs/safer-ffi/0.1.7/safer_ffi/headers/trait.Definer.html), or something along the lines of it (your `GenerationConfig`, with the `Definer` being a property / getter thereof?):

  - something provided as input to all of the `HeaderLanguage` trait functions;
  - which the [`header::Builder`](https://docs.rs/safer-ffi/0.1.7/safer_ffi/headers/struct.Builder.html) could be tweaking

**Comment by pixsperdavid** [comment]
See #226
